internal/controller/storageclaim_controller.go

```diff
@@ -424,6 +464,48 @@ func (r *StorageClaimReconciler) reconcilePhases() (reconcile.Result, error) {
 				if err != nil {
 					return reconcile.Result{}, fmt.Errorf("failed to create or update VolumeSnapshotClass: %s", err)
 				}
+			case "VolumeReplicationClass":
+				data := map[string]string{}
+				err = json.Unmarshal(resource.Data, &data)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to unmarshal StorageClaim configuration response: %v", err)
+				}
+				var volumeReplicationClass, volumeReplicationClassFlattenMode *replicationv1alpha1.VolumeReplicationClass
+				data["replication.storage.openshift.io/replication-secret-namespace"] = r.OperatorNamespace
+				// generate a new clusterID for cephfs subvolumegroup, as
+				// storageclaim is clusterscoped resources using its
+				// hash as the clusterID
+				data["clusterID"] = r.storageClaimHash
+
+				volumeReplicationClass = &replicationv1alpha1.VolumeReplicationClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: r.storageClaim.Name,
+					},
+				}
+				volumeReplicationClassFlattenMode = &replicationv1alpha1.VolumeReplicationClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: fmt.Sprintf("%s-flatten", r.storageClient.Name),
+					},
+				}
+
+				err = utils.CreateOrReplace(r.ctx, r.Client, volumeReplicationClass, func() error {
+					utils.AddLabels(volumeReplicationClass, resource.Labels)
+					utils.AddAnnotation(volumeReplicationClass, storageClaimAnnotation, r.storageClaim.Name)
+					volumeReplicationClass.Spec.Parameters = data
+					return nil
+				})
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to create or update VolumeReplicationClass: %s", err)
+				}
+				err = utils.CreateOrReplace(r.ctx, r.Client, volumeReplicationClassFlattenMode, func() error {
+					utils.AddLabels(volumeReplicationClassFlattenMode, resource.Labels)
+					utils.AddAnnotation(volumeReplicationClassFlattenMode, storageClaimAnnotation, r.storageClaim.Name)
+					volumeReplicationClassFlattenMode.Spec.Parameters = data
+					return nil
+				})
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to create or update VolumeReplicationClass for flatten mode: %s", err)
+				}
```

Comment:
1. Since we have the same code here could we put it in a loop?
2. We are also sending annotations for VRC from providers, I don't see that included here

Reply:
- `CreateOrReplcae` was created to handle generic object that refers to the params field. No need to generalize further than that and on top of that this won't be an ever growing list of resources IMO.
- Added `AddAnnotations` util function